galley/pkg/config/source/kube/apiserver/status/controller.go

```diff
@@ -0,0 +1,146 @@
+	defer c.mu.Unlock()
+
+	if c.state != nil {
+		return
+	}
+	c.state = newState()
+
+	ifaces := make(map[collection.Name]dynamic.NamespaceableResourceInterface)
+	for _, r := range resources {
+		if r.Disabled {
+			continue
+		}
+
+		iface, err := p.GetDynamicResourceInterface(r)
+		if err != nil {
+			scope.Source.Errorf("Unable to create a dynamic resource interface for resource %v", r.CanonicalResourceName())
+		}
+		ifaces[r.Collection.Name] = iface
+	}
+
+	c.wg.Add(1)
+	go run(c.state, ifaces, &c.wg)
+}
+
+// Stop the controller
+func (c *Controller) Stop() {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	if c.state != nil {
+		c.state.quiesceWork()
+		c.wg.Wait()
+		c.state = nil
+	}
+}
+
+// UpdateResourceStatus is called by the source to relay the currently observed status of a resource.
+func (c *Controller) UpdateResourceStatus(
+	col collection.Name, name resource.Name, version resource.Version, status interface{}) {
+	c.state.setLastKnown(col, name, version, status)
+}
+
+// Report the given set of messages towards particular resources.
+func (c *Controller) Report(messages Messages) {
+	c.state.applyMessages(messages)
+}
+
+func run(state *state, ifaces map[collection.Name]dynamic.NamespaceableResourceInterface, wg *sync.WaitGroup) {
+mainloop:
+	for {
+		st, ok := state.dequeueWork()
+		if !ok {
+			break mainloop
+		}
+
+		if !st.needsChange() {
+			// It is possible for a resource to have its status updated twice before being processed in the queue.
+			// The first update can cause the resource to be put in queue, and the subsequent update will change
+			// the state back and cause needsChange to return false. In this case, simply skip this entry.
+			continue mainloop
+		}
+
+		iface := ifaces[st.key.col]
+		if iface == nil {
+			scope.Source.Errorf("No updater available for diagnostic message(s) for '%v/%v'.", st.key.col, st.key.res)
+			continue
+		}
+
+		ns, n := st.key.res.InterpretAsNamespaceAndName()
+		u, err := iface.Namespace(ns).Get(n, metav1.GetOptions{ResourceVersion: string(st.lastKnownVersion)})
+		if err != nil {
+			scope.Source.Errorf("Unable to read the resource while trying to update status: %v(%v): %v",
+				st.key.col, st.key.res, err)
+			continue mainloop
+		}
+
+		// Ensure that the resource we read has the same version as the version for which diagnostic message was
+		// generated for.
+		if st.desiredStatusVersion != resource.Version("") && u.GetResourceVersion() != string(st.desiredStatusVersion) {
+			scope.Source.Debugf("Skipping due to version mismatch: %v(%v): %v !=% v",
+				st.key.col, st.key.res, u.GetResourceVersion(), st.desiredStatusVersion)
+			continue mainloop
+		}
+
+		if st.desiredStatus != nil {
+			u.Object["status"] = st.desiredStatus
+		} else {
+			delete(u.Object, "status")
+		}
+
+		_, err = iface.Namespace(ns).UpdateStatus(u, metav1.UpdateOptions{})
+		if err != nil {
+			// TODO: Reinsert work? It probably makes sense to reinsert (with a delay), in case of a transient failure.
```

Comment:
Not sure what you mean?

Reply:
I had been wondering if UpdateStatus should be responsible for retries instead of throwing them back into the work queue, but after our offline discussion I doubt that makes sense here.